LUN-276: Manage Selection Screen

SteffiStuffel opened the pull request:
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

Handling of loading and errors will be done in LUN-301.
I not also did LUN-276 (adding of a profession) here, but implemented the hole ManageSelectionScreen, so I also did LUN-277 (removing of a profession). These user-story-tickets as a developer task does not really make sense, so I think it is good, we decided to move the "functional"(=fachlich) user stories into the epic and create developer tasks that make sense for developing, e.g. in this case it should have been "Implement Manage Selection Screen"

klinzo commented on src/routes/home/HomeScreen.tsx:
```diff
@@ -36,13 +36,14 @@ interface HomeScreenProps {
   const isCustomDisciplineEmpty = !customDisciplines || customDisciplines.length <= 0
 
   useFocusEffect(
     React.useCallback(() => {
       refreshCustomDisciplines()
-    }, [refreshCustomDisciplines])
+      refreshSelectedProfessions()
+    }, [refreshCustomDisciplines, refreshSelectedProfessions])
   )
```
Optional since not added in this PR: No useCallback should be needed since refreshes are already wrapped in one in the hooks.

SteffiStuffel replied:
Without the callback we get infinite calls. `Please report: Excessive number of pending callbacks: 501. Some pending callbacks that might have leaked by never being called from native code`

klinzo replied:
Perhaps since we have to wrap it in a new arrow function to be able to call both refreshes? Splitting it in two `useFocusEffect`s might help (did not actually test it so not sure). Anyway, fine as it is.

klinzo commented on src/routes/__tests__/ProfessionSelectionScreen.spec.tsx:
```diff
@@ -0,0 +1,102 @@
+    }
+  })
+
+  it('should select profession when clicked', async () => {
```
Minor: I think it would be better to stick to the term `pressed`

klinzo commented on src/routes/manage-disciplines/ManageDisciplinesScreen.tsx:
```diff
@@ -0,0 +1,119 @@
+  } = useReadSelectedProfessions()
+  const { data: customDisciplines, refresh: refreshCustomDisciplines } = useReadCustomDisciplines()
+
+  const refresh = useCallback(() => {
```
No callback should be needed.

klinzo commented on src/routes/manage-disciplines/ManageDisciplinesScreen.tsx:
```diff
@@ -0,0 +1,119 @@
+    refreshSelectedProfessions()
+  }, [refreshCustomDisciplines, refreshSelectedProfessions])
+
+  React.useEffect(() => {
+    refresh()
+    const willFocusSubscription = navigation.addListener('focus', () => {
+      refresh()
+    })
+    return willFocusSubscription
+  }, [navigation, refresh])
```
Any reason why you are not using the focus effect from reactnavigation?
Minor: Unnecessary wrapping of refresh in the listener and no need for a const, can be returned directly.

SteffiStuffel replied:
not sure, why I did this. Tried with FocusEffect and seems to work fine also.

klinzo commented on src/routes/manage-disciplines/ManageDisciplinesScreen.tsx:
```diff
@@ -0,0 +1,119 @@
+    const unselectProfessionAndRefresh = (item: Discipline) => {
+      removeSelectedProfession(item)
+        .then(refreshSelectedProfessions)
+        .catch(err => reportError(err))
```
```suggestion
        .catch(reportError)
```

klinzo commented on src/routes/manage-disciplines/ManageDisciplinesScreen.tsx:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  const navigateToProfessionSelection = () => {
+    navigation.push('Intro', { initialSelection: false })
```
Same below in `navigateToAddCustomDiscipline`
```suggestion
    navigation.navigate('Intro', { initialSelection: false })
```

klinzo reviewed (changes requested):
Tested on android. Works perfectly, really nice work again!

Things noticed:
1. The touch target for the header menu seems a little bit small
2. `Beruf hinzufügen` is centered while `Bereich Hinzufügen` seems to be left aligned
3. The cards on the home screen are quite large such that only the first two are visible on my (not really small phone), maybe we could reduce some spacings here?

SteffiStuffel commented:
> 1. The touch target for the header menu seems a little bit small

Fix it.

> 2. `Beruf hinzufügen` is centered while `Bereich Hinzufügen` seems to be left aligned

Good catch, wasn't like in the designs. Fixed it.

> 3. The cards on the home screen are quite large such that only the first two are visible on my (not really small phone), maybe we could reduce some spacings here?

Removed the spacing a bit here, if we want it smaller, I think we should talk to Nikola first.

SteffiStuffel commented on src/components/AddElement.tsx:
```diff
@@ -0,0 +1,49 @@
+`
+
+const Explanation = styled(ContentSecondary)`
+  padding: 0 ${props => props.theme.spacings.xxl} 0 15%;
```
Yes right. I was really racking my brain here, as we need a padding that has the width of the icon plus the padding. In css i would have used grid, but this is not supported in react native, also css-calc function is not supported. Therefore I decided this is the best solution, although I am also not really happy with it. If you have a better solution, let me know. Otherwise I think it is fine to keep it.

klinzo replied:
I definitely don't have a better solution, css and styling is not really my sahneseite :D maybe @f1sh1918 has one though

SteffiStuffel commented on src/components/AddElement.tsx:
```diff
@@ -0,0 +1,49 @@
+const AddElement = ({ onPress, label, explanation }: Props): ReactElement => (
+  <Root onPress={onPress}>
+    <FlexRow>
+      <Icon width={wp('8%')} height={wp('8%')} testID='add-icon' />
+      <AddCustomDisciplineText>{label}</AddCustomDisciplineText>
```
Though it would make sense here, to check if the component shows the icon, but not really needed. Removed it and replaced it with label.

SteffiStuffel commented on src/components/HorizontalLine.tsx:
```diff
@@ -0,0 +1,13 @@
+  border: solid 1px ${props => props.theme.colors.disabled};
+`
+
+const HorizontalLine = (): ReactElement => <Line />
+
+export default HorizontalLine
```
No, no need. Changed it.

SteffiStuffel commented on src/components/ListItem.tsx:
```diff
@@ -138,17 +146,19 @@ const ListItem = ({
       pressed={pressed}
-      delayLongPress={200}>
+      delayLongPress={200}
+      testID='list-item'>
```
hmh, I was orienting myself on the tests above. I added to the test, that if disabled, the onPress-function is not called, but kept the test for correct background-color, as I am not really sure, if we should remove it.

klinzo replied:
Yeah then it's fine, lets keep it :)

SteffiStuffel commented on src/navigation/Navigator.tsx:
```diff
@@ -118,13 +120,18 @@ const Navigator = (): JSX.Element | null => {
+          options={{ headerShown: false }}
+          name='Intro'
+          component={IntroScreen}
+          initialParams={{ initialSelection: true }}
```
Something like Scope- or SectorSelection?

klinzo replied:
We already have a `DisciplineSelection`, right? TBH I don't have an overview which is used when anymore and about the different implementations. Maybe we would be able to merge some stuff? Of course not in the scope of this PR. I am fine `ScopeSelection` I guess, not really nice but okay.

SteffiStuffel replied:
I already created a ticket for it [LUN-300](https://issues.tuerantuer.org/browse/LUN-300).

We have ScopeSelection and ProfessionSelection for selecting a profession that should be learnt. And we have DisciplineSeleciton for selection which discipline of a profession should be learnt.

klinzo replied:
Works for me, renaming can be done there than as well :+1:

SteffiStuffel commented on src/routes/ProfessionSelectionScreen.tsx:
```diff
@@ -50,26 +50,36 @@ const ProfessionSelectionScreen = ({ route, navigation }: ProfessionSelectionScr
     } else {
-      await AsyncStorage.pushSelectedProfession(selectedItem).then(refreshSelectedProfessions)
+      await AsyncStorage.pushSelectedProfession(selectedItem).then(() => {
+        if (!initialSelection) {
+          navigation.navigate('ManageDisciplines')
+        } else {
+          refreshSelectedProfessions()
+        }
```
Yes please.

SteffiStuffel commented on src/routes/add-custom-discipline/AddCustomDisciplineScreen.tsx:
```diff
@@ -90,7 +90,7 @@ const AddCustomDiscipline = ({ navigation }: AddCustomDisciplineScreenProps): JS
     loadGroupInfo(code)
       .then(async () => AsyncStorage.setCustomDisciplines([...customDisciplines, code]))
-      .then(() => navigation.navigate('Home'))
+      .then(() => navigation.goBack())
```
yes please

f1sh1918 commented on src/components/AddElement.tsx:
```diff
@@ -0,0 +1,44 @@
+`
+
+const Explanation = styled(ContentSecondary)`
+  padding: 0 ${props => props.theme.spacings.xxl} 0 11%;
```
can't we use either xl(10%) or xxl(12%) ? even if the design differs we should use defined spacings

SteffiStuffel replied:
this padding needs to be the same as the width of the icon plus the padding of the text, so these text is exactly beneath the other one. For better understanding have a look at the designs, which makes it clear.

f1sh1918 replied:
alright keep it like that.

f1sh1918 commented on src/components/ListItem.tsx:
```diff
@@ -138,17 +146,19 @@ const ListItem = ({
         <DescriptionContainer>
           {badgeLabel && <BadgeLabel pressed={pressed}>{badgeLabel}</BadgeLabel>}
-          <Description pressed={pressed}>{description}</Description>
+          {description && description.length > 0 && <Description pressed={pressed}>{description}</Description>}
```
```suggestion
           {description?.length > 0 && <Description pressed={pressed}>{description}</Description>}
```
you can use shorthand

f1sh1918 commented on src/components/ListItem.tsx:
```diff
@@ -138,17 +146,19 @@ const ListItem = ({
       <FlexContainer>
         {titleToRender}
         <DescriptionContainer>
           {badgeLabel && <BadgeLabel pressed={pressed}>{badgeLabel}</BadgeLabel>}
```
```suggestion
           {badgeLabel && badgeLabel.length > 0 && <BadgeLabel pressed={pressed}>{badgeLabel}</BadgeLabel>}
```
optional since you didn't change that line

f1sh1918 commented on src/components/ListItem.tsx:
```diff
@@ -16,9 +15,14 @@ export const GenericListItemContainer = styled.Pressable`
+    return prop.pressed ? prop.theme.colors.primary : prop.theme.colors.backgroundAccent
+  }};
   border: 1px solid ${prop => (prop.pressed ? prop.theme.colors.primary : prop.theme.colors.disabled)};
   padding: ${props =>
```
optional because not touched:
`padding: ${props => props.theme.spacings.sm};`
not sure why padding left & right are not equal

SteffiStuffel replied:
Because we use this list item also for the selection of a module and there we have this ">" icon, which should be more on the right. https://app.zeplin.io/project/5fc7c98a435298126bcfd4ee/screen/5fd7ad2c3cfec77f5c68f10c

f1sh1918 commented on src/routes/manage-selections/ManageSelectionsScreen.tsx:
```diff
@@ -0,0 +1,112 @@
+        label={labels.home.addCustomDiscipline}
+        explanation={labels.manageDisciplines.descriptionAddCustomDiscipline}
+      />
+      <Padding />
```
can't find any difference without this view container. whats the purpose?

SteffiStuffel replied:
If you add 5 professions the last element is close to the bottom.

f1sh1918 commented on src/components/HeaderWithMenu.tsx:
```diff
@@ -22,25 +26,44 @@ const SmileIconStyle = styled.Pressable`
 `
 
-const MenuIconContainer = styled.Pressable`
+const HeaderButtonsContainer = styled.View`
```
you can just use flexbox instead of positioning absolute

```js
const HeaderButtonsContainer = styled.View`
  align-self: flex-end;
`
```

SteffiStuffel replied:
nice

f1sh1918 reviewed (commented):
Works fine on iOS just found some minor styling issues/improvements

- Red circle of delete icon in `ManageSelectionScreen` is missing
- I think we could remove the disabled color from ListItem (check code comment)
- Design of `ManageSelectionScreen`does not fit with the implementation

SteffiStuffel commented:
> Works fine on iOS just found some minor styling issues/improvements
> 
> * Red circle of delete icon in `ManageSelectionScreen` is missing
> * I think we could remove the disabled color from ListItem (check code comment)
> * Design of `ManageSelectionScreen`does not fit with the implementation

You can see the version, I implemented here, won't do the changes on the screen now, but fixed the red circle icon
https://app.zeplin.io/project/5fc7c98a435298126bcfd4ee/screen/62419215eaccb511191ee013/version/62419215eaccb511191ee014

Fixed disable color.

klinzo commented on src/components/ListItem.tsx:
```diff
@@ -98,7 +98,9 @@ const ListItem = ({
 }: ListItemProps): ReactElement => {
   const [pressInY, setPressInY] = useState<number | null>(null)
-  const [pressed, setPressed] = useState<boolean>(false)
+  const [pressed, setPressedState] = useState<boolean>(false)
+  const setPressed = (pressed: boolean): void => onPress && setPressedState(pressed)
```
Please find another name for current `setPressed` and rename the `useState` method to `setPressed`. I think this is a convention we should stick to that `useState` value and setter are always the same. How about or similar:
```suggestion
  const [pressed, setPressed] = useState<boolean>(false)
  const updatePressed = (pressed: boolean): void => onPress && setPressedState(pressed)
```
I also think it should depend on `disabled` instead of on `onPress`. If the button is not disabled it should react to user presses even if there is no `onPress` method. At least that is what I'd expect.

SteffiStuffel replied:
My thoughts about this were:
1) you don't have to think that setPress needs different handling and just call it as normal and side-case handling is done for you
2) disabling changes the color too, and we what three types of behaviour: a) disabled = grey and no action on click b) disabled = not-grey but no action on click c) not disabled = not-grey and action on click.

f1sh1918 reviewed (changes requested):
![image](https://user-images.githubusercontent.com/37902063/164189236-aed67dda-44a1-45cc-a62a-49e4c0cb3b1e.png)

In `ManageSelectionsScreen`and `CustomDisciplineItem` different paddings for CloseIconRed Pressable Container are used
